From 5b776c59119ee2a391e5c57e7b5350327f399916 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:24:29 -0400
Subject: [tools] perf/trace: Beautify eventfd2 'flags' arg

Message-id: <1408462094-14194-101-git-send-email-jolsa@redhat.com>
Patchwork-id: 88076
O-Subject: [PATCH RHEL7.1 BZ1131394 100/325] perf trace: Beautify eventfd2 'flags' arg
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 49af9e93adfa11d50435aa079299a765843532fc
Author: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Thu Sep 12 12:18:56 2013 -0300

description
===========
 61.168 ( 0.004 ms): 24267 eventfd2(flags: CLOEXEC|NONBLOCK) = 9
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/builtin-trace.c b/tools/perf/builtin-trace.c
index f41254f..53d1e3f 100644
--- a/tools/perf/builtin-trace.c
+++ b/tools/perf/builtin-trace.c
@@ -13,6 +13,7 @@
 
 #include <libaudit.h>
 #include <stdlib.h>
+#include <sys/eventfd.h>
 #include <sys/mman.h>
 #include <linux/futex.h>
 
@@ -456,6 +457,32 @@ static size_t syscall_arg__scnprintf_open_flags(char *bf, size_t size,
 
 #define SCA_OPEN_FLAGS syscall_arg__scnprintf_open_flags
 
+static size_t syscall_arg__scnprintf_eventfd_flags(char *bf, size_t size,
+         struct syscall_arg *arg)
+{
+ int printed = 0, flags = arg->val;
+
+ if (flags == 0)
+  return scnprintf(bf, size, "NONE");
+#define P_FLAG(n) \
+ if (flags & EFD_##n) { \
+  printed += scnprintf(bf + printed, size - printed, "%s%s", printed ? "|" : "", #n); \
+  flags &= ~EFD_##n; \
+ }
+
+ P_FLAG(SEMAPHORE);
+ P_FLAG(CLOEXEC);
+ P_FLAG(NONBLOCK);
+#undef P_FLAG
+
+ if (flags)
+  printed += scnprintf(bf + printed, size - printed, "%s%#x", printed ? "|" : "", flags);
+
+ return printed;
+}
+
+#define SCA_EFD_FLAGS syscall_arg__scnprintf_eventfd_flags
+
 static size_t syscall_arg__scnprintf_signum(char *bf, size_t size, struct syscall_arg *arg)
 {
  int sig = arg->val;
@@ -516,6 +543,8 @@ static struct syscall_fmt {
  { .name     = "brk",     .hexret = true,
    .arg_scnprintf = { [0] = SCA_HEX, /* brk */ }, },
  { .name     = "connect",    .errmsg = true, },
+ { .name     = "eventfd2",   .errmsg = true,
+   .arg_scnprintf = { [1] = SCA_EFD_FLAGS, /* flags */ }, },
  { .name     = "fcntl",     .errmsg = true,
    .arg_scnprintf = { [1] = SCA_STRARRAY, /* cmd */ },
    .arg_parm  = { [1] = &strarray__fcntl_cmds, /* cmd */ }, },
-- 
1.7.1